File: shared/ios/ReleafCore/Sources/ReleafCoreScan/BusinessCard/Classifier/EmailClassifier.swift
/*
 * EmailClassifier.swift  — mirror of EmailClassifier.kt
 */

import Foundation

public struct EmailClassifier: Classifier {

    public init() {}

    public func classify(layout: LayoutContext, weights: ScoringWeights) -> [FieldCandidate] {
        var out: [FieldCandidate] = []
        for (idx, block) in layout.blocks.enumerated() {
            let nsText = block.text as NSString
            let matches = Self.regex.matches(
                in: block.text,
                range: NSRange(location: 0, length: nsText.length)
            )
            for m in matches {
                let raw = nsText.substring(with: m.range)
                let score = weights.emailBase + engineConfidenceBonus(block, weights)
                out.append(FieldCandidate(
                    sourceBlockIndex: idx,
                    text:             raw.lowercased().trimmingCharacters(in: .whitespaces),
                    kind:             .email,
                    score:            score
                ))
            }
        }
        return out
    }

    // Pragmatic email regex — accepts the practical subset that
    // appears on cards. Not RFC-strict.
    private static let regex: NSRegularExpression = {
        let pattern = #"[A-Za-z0-9](?:[A-Za-z0-9._%+\-]*[A-Za-z0-9])?@[A-Za-z0-9](?:[A-Za-z0-9.\-]*[A-Za-z0-9])?\.[A-Za-z]{2,24}"#
        // swiftlint:disable:next force_try
        return try! NSRegularExpression(pattern: pattern)
    }()
}

File: shared/android/shared/scan/src/main/kotlin/app/releaf/shared/scan/businesscard/classifier/EmailClassifier.kt
/*
 * EmailClassifier.kt
 *
 * Identifies blocks that contain an email address. Regex hit alone
 * is high-signal (false-positive rate is near-zero on real cards),
 * so we lean heavily on the base score and add only a small engine-
 * confidence bonus on top.
 *
 * RFC 5322 in full is overkill here — the tighter pattern below
 * accepts the practical subset that appears on cards (printed
 * emails). One classifier emits one candidate per match; multi-
 * email blocks (rare) emit one per address.
 */

package app.releaf.shared.scan.businesscard.classifier

import app.releaf.shared.scan.businesscard.FieldCandidate
import app.releaf.shared.scan.businesscard.FieldKind
import app.releaf.shared.scan.businesscard.ScoringWeights
import app.releaf.shared.scan.businesscard.pipeline.LayoutContext
import app.releaf.shared.scan.businesscard.pipeline.engineConfidenceBonus

class EmailClassifier : Classifier {

    override fun classify(layout: LayoutContext, weights: ScoringWeights): List<FieldCandidate> {
        val out = mutableListOf<FieldCandidate>()
        for ((idx, block) in layout.blocks.withIndex()) {
            for (match in EMAIL_REGEX.findAll(block.text)) {
                val score = weights.emailBase + engineConfidenceBonus(block, weights)
                out += FieldCandidate(
                    sourceBlockIndex = idx,
                    text             = match.value.lowercase().trim(),
                    kind             = FieldKind.EMAIL,
                    score            = score,
                )
            }
        }
        return out
    }

    companion object {
        // Pragmatic email regex. Accepts standard local-part + domain
        // with TLD; rejects unbalanced punctuation. Not RFC-strict.
        private val EMAIL_REGEX = Regex(
            """[A-Za-z0-9](?:[A-Za-z0-9._%+\-]*[A-Za-z0-9])?@[A-Za-z0-9](?:[A-Za-z0-9.\-]*[A-Za-z0-9])?\.[A-Za-z]{2,24}"""
        )
    }
}
